gensim/models/phrases.py

```diff
@@ -554,237 +610,73 @@ def add_vocab(self, sentences):
             self.min_reduce = max(self.min_reduce, min_reduce)
             for word, count in vocab.items():
                 self.vocab[word] += count
             if len(self.vocab) > self.max_vocab_size:
                 utils.prune_vocab(self.vocab, self.min_reduce)
                 self.min_reduce += 1
-            logger.info("merged %s", self)
         else:
-            # in common case, avoid doubling gigantic dict
-            logger.info("using %i counts as vocab in %s", len(vocab), self)
+            # Optimization for a common case: the current vocab is empty, so apply
+            # the new vocab directly, no need to double it in memory.
             self.vocab = vocab
-
-    def export_phrases(self, sentences, out_delimiter=b' ', as_tuples=False):
-        """Get all phrases that appear in 'sentences' that pass the bigram threshold.
-
-        Parameters
-        ----------
-        sentences : iterable of list of str
-            Text corpus.
-        out_delimiter : str, optional
-            Delimiter used to "glue" together words that form a bigram phrase.
-        as_tuples : bool, optional
-            Yield `(tuple(words), score)` instead of `(out_delimiter.join(words), score)`?
-
-        Yields
-        ------
-        ((str, str), float) **or** (str, float)
-            Phrases detected in `sentences`. Return type depends on the `as_tuples` parameter.
-
-        Example
-        -------
-        .. sourcecode:: pycon
-
-            >>> from gensim.test.utils import datapath
-            >>> from gensim.models.word2vec import Text8Corpus
-            >>> from gensim.models.phrases import Phrases
-            >>>
-            >>> sentences = Text8Corpus(datapath('testcorpus.txt'))
-            >>> phrases = Phrases(sentences, min_count=1, threshold=0.1)
-            >>>
-            >>> for phrase, score in phrases.export_phrases(sentences):
-            ...     pass
-
-        """
-        analyze_sentence = functools.partial(
-            self.analyze_sentence,
-            threshold=self.threshold,
-            common_terms=self.common_terms,
-            scorer=functools.partial(
-                self.scoring,
-                len_vocab=float(len(self.vocab)),
-                min_count=float(self.min_count),
-                corpus_word_count=float(self.corpus_word_count),
-            ),
+        logger.info("merged %s", self)
+
+    def score_candidate(self, word_a, word_b, in_between):
+        # Micro optimization: check for quick early-out conditions, before the actual scoring.
+        word_a_cnt = self.vocab[word_a]
+        if word_a_cnt <= 0:
+            return None, None
+
+        word_b_cnt = self.vocab[word_b]
+        if word_b_cnt <= 0:
+            return None, None
+
+        phrase = self.delimiter.join([word_a] + in_between + [word_b])
+        # XXX: Why do we care about *all* phrase tokens? Why not just score the start+end bigram?
```

Comment:
@alexgarel @gojomo do you remember why we score on `word_a + in_between + word_b`, instead of just `word_a word_b`? What's the point of including the common words in the calculation?

Reply:
I'd guess that tracks back to feature #1258, with an abortive implementation in #1567, arriving in #1568. It sounded useful to me in some situations, but hard enough to understand, and twisty enough in its effect on existing code, that I suggested a couple times it be a separate class, rather than options on `Phrases`.

Reply:
(And relatedly: my hunch is this has no effect unless that non-default option is activated, because `in_between` is then always empty.)

Reply:
Your hunch is correct – `common_words` is empty by default.

I'm leaving the logic of `#['eye', 'of', 'the', 'beholder']` vs `#['eye', 'beholder']` unchanged. I don't know if anyone's using `common_words` at all, but if they are, let's avoid surprises.

Reply:
Re: defaults

Despite my own perspective as an English monglot, I'd hate to change a language-agnostic algorithm, that previously had no English-specific initialization, to become English-specific – even in a major version bump, with lots of warning. (Providing a handy switch/presets that makes enabling an English-specific mode, and/or other languages, sounds useful though.)

Reply:
Addressed in #2979 .